src/carousel/index.ts

```diff
@@ -0,0 +1,206 @@
+
+    const posterior = this.props.items
+      .slice(0, numCloned)
+      .map((data, index) => ({ ...data }));
+
+    return prior.concat(this.props.items).concat(posterior);
+  }
+
+  slideHandler = (slide: number) => {
+    const { speed } = this.props;
+    const from = this.state.currentSlide;
+    const length = this.props.items.length;
+    const onEdge = slide >= length || slide <= 0;
+
+    // don't need to consider on edge if speed is 0;
+    if (!speed) {
+      this.resetCurrentSlideNum(from, slide, length);
+      return;
+    }
+
+    if (!(this.state.transitioning && onEdge)) {
+      // make the transition
+      this.set({ currentSlide: slide, transitioning: true });
+
+      if (onEdge) {
+        // reset to non-cloned slide
+        // tslint:disable-next-line:max-line-length
+        this.animationEndCallback = setTimeout(() => this.resetToRealSlide(from, slide, length), speed);
+      } else {
+        this.refs.track.addEventListener('transitionend', this.disableTransition, false);
+      }
+    }
+  }
+
+  resetToRealSlide = (from: number, to: number, length: number) => {
+    this.state.transitioning = false;
+    this.resetCurrentSlideNum(from, to, length);
+    this.animationEndCallback = null;
+  }
+
+  resetCurrentSlideNum = (from: number, to: number, length: number) => {
+    const currentSlide = from < to ? to % length : (to % length + length) % length;
+    this.set({ currentSlide });
+  }
+
+  disableTransition = () => {
+    this.refs.track.removeEventListener('transitionend', this.disableTransition, false);
+    this.set({ transitioning: false });
+  }
+
+  trackStyle = () => {
+    const { speed } = this.props;
+    if (this.state.transitioning && speed) {
+      const transition = `${speed}ms ease`;
+      return {
+        ...this.getStaticTrackStyle(),
+        '-webkit-transition': transition,
+        transition,
+        '-ms-transition': transition
+      };
+    } else {
+      return this.getStaticTrackStyle();
+    }
+  }
+
+  getStaticTrackStyle = () => {
+    const slideWidth = this.getSlideWidth();
+    const slideCount = this.props.items.length;
+    const slidestoshow = this.props.slidestoshow;
+    const trackWidth = (slideCount + 4 * slidestoshow - 2) * slideWidth;
+    const pos = Carousel.calculatePosition(this.state.currentSlide, slideWidth, slidestoshow);
+    const transform = `translateX(-${pos}px)`;
+
+    return {
+      width: `${trackWidth}px`,
+      transform,
+      '-webkit-transform': transform,
+      '-ms-transform': transform
+    };
+  }
+
+  getSlideWidth = () => this.refs.wrapper.offsetWidth / this.props.slidestoshow;
+
+  onTouchStart = (e: TouchEvent & Carousel.Event) => {
+    this.state.touchObject = {
+      startX: e.touches[0].pageX,
+      startY: e.touches[0].pageY
+    };
+
+    e.target.addEventListener('touchend', this.onTouchEnd);
+  }
+
+  onTouchEnd = (e: TouchEvent & Carousel.Event) => {
```

Comment:
could do the same here with the method signature
```js
onTouchEnd = ({ changedTouches: [{ pageX, pageY }], target }: TouchEvent & Carousel.Event) 
```

Reply:
TouchList doesn't allow this deconstruct